jscomp/test/Import.res

```diff
@@ -1,18 +1,12 @@
-let each = Js.import(Belt.List.forEach)
-
-let eachInt = (list: list<int>, f: int => unit) =>
-  Js.Promise.then_(each => list->each(f)->Js.Promise.resolve, each)
-
-module type BeltList = module type of Belt.List
-
-let beltAsModule = Js.import(module(Belt.List: BeltList))
-
 let eachIntAsync = async (list: list<int>, f: int => unit) => {
-  let each = await each
-  list->each(f)
+  list->(await Js.import(Belt.List.forEach))(f)
 }
 
-let eachLazy = () => Js.import(Belt.List.forEach)
-
 let eachIntLazy = (list: list<int>, f: int => unit) =>
-  eachLazy() |> Js.Promise.then_(each => list->each(f)->Js.Promise.resolve)
+  Js.import(Belt.List.forEach) |> Js.Promise.then_(each => list->each(f)->Js.Promise.resolve)
+
+let _ = list{1, 2, 3}->eachIntLazy(n => Js.log2("lazy", n))
+let _ = list{1, 2, 3}->eachIntAsync(n => Js.log2("async", n))
+
+module type BeltList = module type of Belt.List
+let beltAsModule = Js.import(module(Belt.List: BeltList))
```

Comment:
@mattdamon108 

Since this encoding is going to be painful to use, as discussed, what is an alternative encoding to represent:
```res
module M = await import(Belt)
```
How about this:

```res
module M = @res.await Belt
```
that's already the way normal `await` for values is currently represented internally.

Actual syntax sugar to be added later.

Reply:
Regarding the module mechanism
```rescript
module M = @res.await Belt.List // 1
let each = M.forEach // 2
```
What would the js output be?
```js
let each = import("...").then(m => m.forEach)
```
If so, how about without expression 2?
```js
let m = import("...") // ??
```
Without the expression 2, only module binding generating the value seems a little weird to me. What do you think?

Reply:
I think we can dynamic import without the module mechanism:
```rescript
let default = Js.import(Comp.default)
```

Reply:
> Regarding the module mechanism
> 
> ```rescript
> module M = @res.await Belt.List // 1
> let each = M.forEach // 2
> ```
> 
> What would the js output be?
> 
> ```js
> let each = import("...").then(m => m.forEach)
> ```
> 
> If so, how about without expression 2?
> 
> ```js
> let m = import("...") // ??
> ```
> 
> Without the expression 2, only module binding generating the value seems a little weird to me. What do you think?

Line 2 is another line. The mechanism should operate only on line 1.

Reply:
What code is generated for `module M = Belt.List`?
It should be a variation of that same code, but dynamic.

Reply:
> I think we can dynamic import without the module mechanism:
> 
> ```rescript
> let default = Js.import(Comp.default)
> ```

It's not about components, it's about modules in general.

Reply:
> What code is generated for `module M = Belt.List`? It should be a variation of that same code, but dynamic.

Until any value from `M`, the js output is just:
```js
var M;
```

Reply:
```rescript
module M = @res.await Belt.List
let each = M.forEach // 2
```
If the user writes the line 2, then line 2 needs to be compiled to js as dynamic import. I guess we can use `res.await` in Typedtree.attributes in `Belt.List`

Reply:
Surely the first line is not compiler to nothing, just does not emit code per se.
It must be compiled to something ready to emit code when used.
And I guess it should be compiled to something ready to emit different code (the dynamic one).
Without looking at line 2.

Reply:
So I think there should be line `3` and see what additional work gets done to produce line 3:
```res
let foo = M.foo // 3
```

Reply:
@cristianoc ⬆️  Is it what you intended?

Reply:
> @cristianoc ⬆️ Is it what you intended?

That looks great